src/Shouldly/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs

```diff
@@ -102,62 +103,62 @@ public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Ex
             ShouldAllBe(actual, elementPredicate, () => null);
         }
 
-        public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, string customMessage)
+        public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, string? customMessage)
         {
             ShouldAllBe(actual, elementPredicate, () => customMessage);
         }
 
-        public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, [InstantHandle] Func<string> customMessage)
+        public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Expression<Func<T, bool>> elementPredicate, [InstantHandle] Func<string?>? customMessage)
         {
             var condition = elementPredicate.Compile();
             var actualResults = actual.Where(part => !condition(part));
             if (actualResults.Any())
                 throw new ShouldAssertException(new ActualFilteredWithPredicateShouldlyMessage(elementPredicate.Body, actualResults, actual, customMessage).ToString());
         }
 
-        public static void ShouldBeEmpty<T>(this IEnumerable<T> actual)
+        public static void ShouldBeEmpty<T>([NotNull] this IEnumerable<T>? actual)
```

Comment:
`[NotNull]` initially seems to conflict with the `?`, but there is no conflict. `[NotNull]` is indicating that if this call returns, the expression that had been passed to `actual` cannot have been null. At the same time, the method accepts null inputs without a compile-time warning.

You could also choose to go with `ShouldBeEmpty<T>(this IEnumerable<T> actual)` in which case the compiler would warn if the user does `foo.ShouldBeEmpty();` without first proving to the compiler that `foo` can't be null (or using `!`).